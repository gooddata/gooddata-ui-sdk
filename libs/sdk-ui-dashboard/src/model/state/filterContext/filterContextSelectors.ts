// (C) 2021 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import { DashboardState } from "../types";
import invariant from "ts-invariant";
import {
    FilterContextItem,
    IDashboardAttributeFilter,
    IDashboardDateFilter,
    isDashboardAttributeFilter,
    isDashboardDateFilter,
} from "@gooddata/sdk-backend-spi";
import { areObjRefsEqual, ObjRef } from "@gooddata/sdk-model";

const selectSelf = createSelector(
    (state: DashboardState) => state,
    (state) => state.filterContext,
);

/**
 * This selector returns dashboard's filter context. It is expected that the selector is called only after the filter
 * context state is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @internal
 */
export const selectFilterContext = createSelector(selectSelf, (filterContextState) => {
    invariant(filterContextState.filterContext, "attempting to access uninitialized filter context state");

    return filterContextState.filterContext!;
});

/**
 * This selector returns dashboard's filter context filters. It is expected that the selector is called only after the filter
 * context state is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @internal
 */
export const selectFilterContextFilters = createSelector(
    selectFilterContext,
    (filterContext): FilterContextItem[] => filterContext.filters,
);

/**
 * This selector returns dashboard's filter context attribute filters. It is expected that the selector is called only after the filter
 * context state is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @internal
 */
export const selectFilterContextAttributeFilters = createSelector(
    selectFilterContextFilters,
    (filters): IDashboardAttributeFilter[] => filters.filter(isDashboardAttributeFilter),
);

/**
 * This selector returns dashboard's filter context date filter. It is expected that the selector is called only after the filter
 * context state is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @internal
 */
export const selectFilterContextDateFilter = createSelector(selectFilterContextFilters, (filters):
    | IDashboardDateFilter
    | undefined => filters.find(isDashboardDateFilter));

/**
 * Creates a selector for selecting attribute filter by its displayForm {@link @gooddata/sdk-model#ObjRef}.
 *
 * @internal
 */
export const makeSelectFilterContextAttributeFilterByDisplayForm = () =>
    createSelector(
        selectFilterContextAttributeFilters,
        (_: any, displayForm: ObjRef) => displayForm,
        (attributeFilters, displayForm) =>
            attributeFilters.find((filter) =>
                areObjRefsEqual(filter.attributeFilter.displayForm, displayForm),
            ),
    );

/**
 * Creates a selector for selecting attribute filter by its localId.
 *
 * @internal
 */
export const makeSelectFilterContextAttributeFilterByLocalId = () =>
    createSelector(
        selectFilterContextAttributeFilters,
        (_: any, localId: string) => localId,
        (attributeFilters, localId) =>
            attributeFilters.find((filter) => filter.attributeFilter.localIdentifier === localId),
    );

/**
 * Creates a selector for selecting attribute filter index by its localId.
 *
 * @internal
 */
export const makeSelectFilterContextAttributeFilterIndexByLocalId = () =>
    createSelector(
        selectFilterContextAttributeFilters,
        (_: any, localId: string) => localId,
        (attributeFilters, localId) =>
            attributeFilters.findIndex((filter) => filter.attributeFilter.localIdentifier === localId),
    );
