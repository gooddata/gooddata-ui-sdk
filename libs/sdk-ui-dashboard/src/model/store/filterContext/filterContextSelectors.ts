// (C) 2021-2024 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import { DashboardSelector, DashboardState } from "../types.js";
import { invariant } from "ts-invariant";
import {
    areObjRefsEqual,
    ObjRef,
    FilterContextItem,
    IDashboardAttributeFilter,
    IDashboardDateFilter,
    isDashboardAttributeFilter,
    uriRef,
    idRef,
    IAttributeDisplayFormMetadataObject,
    IFilterContextDefinition,
    IDashboardObjectIdentity,
    isDashboardDateFilterWithDimension,
    isObjRef,
    isDashboardCommonDateFilter,
} from "@gooddata/sdk-model";
import { ObjRefMap, newDisplayFormMap } from "../../../_staging/metadata/objRefMap.js";
import { createMemoizedSelector } from "../_infra/selectors.js";
import compact from "lodash/compact.js";
import isEmpty from "lodash/isEmpty.js";
import { selectSupportsCircularDependencyInFilters } from "../backendCapabilities/backendCapabilitiesSelectors.js";

const selectSelf = createSelector(
    (state: DashboardState) => state,
    (state) => state.filterContext,
);

/**
 * This selector returns original (stored) dashboard's filter context definition.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @returns {@link @gooddata/sdk-backend-spi#IFilterContextDefinition} or `undefined` if original filter context definition is not set.
 *
 * @public
 */
export const selectOriginalFilterContextDefinition: DashboardSelector<IFilterContextDefinition | undefined> =
    createSelector(selectSelf, (filterContextState) => {
        invariant(
            filterContextState.filterContextDefinition,
            "attempting to access uninitialized filter context state",
        );

        return filterContextState.originalFilterContextDefinition;
    });

/**
 * This selector returns original (stored) dashboard's filters.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @returns an array of {@link @gooddata/sdk-backend-spi#FilterContextItem} or an empty array.
 *
 * @public
 */
export const selectOriginalFilterContextFilters: DashboardSelector<FilterContextItem[]> = createSelector(
    selectOriginalFilterContextDefinition,
    (filterContext): FilterContextItem[] => filterContext?.filters ?? [],
);

/**
 * This selector returns current dashboard's filter context definition.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @returns a {@link @gooddata/sdk-backend-spi#IFilterContextDefinition}
 *
 * @public
 */
export const selectFilterContextDefinition: DashboardSelector<IFilterContextDefinition> = createSelector(
    selectSelf,
    (filterContextState) => {
        invariant(
            filterContextState.filterContextDefinition,
            "attempting to access uninitialized filter context state",
        );

        return filterContextState.filterContextDefinition!;
    },
);

/**
 * Selects dashboard's filter context identity.
 *
 * @remarks
 * The identity may be undefined in two circumstances:
 *
 * -  a new, yet unsaved dashboard; the filter context is saved together with the dashboard and after the
 *    save the identity will be known and added
 *
 * -  export of an existing, saved dashboard; during the export the dashboard receives a temporary
 *    filter context that represents values of filters at the time the export was initiated - which may
 *    be different from what is saved in the filter context itself. that temporary context is not
 *    persistent and lives only for that particular export operation.
 *
 * Invocations before initialization lead to invariant errors.
 *
 * @returns a {@link @gooddata/sdk-backend-spi#IDashboardObjectIdentity} or undefined, if the filter context identity is not set.
 *
 * @internal
 */
export const selectFilterContextIdentity: DashboardSelector<IDashboardObjectIdentity | undefined> =
    createSelector(selectSelf, (filterContextState) => {
        // this is intentional; want to fail fast when trying to access an optional identity of filter context \
        // but there is actually no filter context initialized for the dashboard
        invariant(
            filterContextState.filterContextDefinition,
            "attempting to access uninitialized filter context state",
        );

        return filterContextState.filterContextIdentity;
    });

/**
 * Selects list of display form metadata objects referenced by attribute filters.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @returns an array of {@link @gooddata/sdk-backend-spi#IAttributeDisplayFormMetadataObject}
 *
 * @public
 */
export const selectAttributeFilterDisplayForms: DashboardSelector<IAttributeDisplayFormMetadataObject[]> =
    createSelector(selectSelf, (filterContextState) => {
        invariant(
            filterContextState.attributeFilterDisplayForms,
            "attempting to access uninitialized filter context state",
        );

        return filterContextState.attributeFilterDisplayForms;
    });

/**
 * Selects map of display form metadata objects referenced by attribute filters.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @returns a {@link ObjRefMap} of {@link @gooddata/sdk-backend-spi#IAttributeDisplayFormMetadataObject}
 *
 * @internal
 */
export const selectAttributeFilterDisplayFormsMap: DashboardSelector<
    ObjRefMap<IAttributeDisplayFormMetadataObject>
> = createSelector(selectSelf, (filterContextState) => {
    invariant(
        filterContextState.attributeFilterDisplayForms,
        "attempting to access uninitialized filter context state",
    );

    return newDisplayFormMap(filterContextState.attributeFilterDisplayForms);
});

/**
 * This selector returns dashboard's filter context filters.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextFilters: DashboardSelector<FilterContextItem[]> = createSelector(
    selectFilterContextDefinition,
    (filterContext): FilterContextItem[] => filterContext.filters,
);

/**
 * This selector returns dashboard's filter context attribute filters.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextAttributeFilters: DashboardSelector<IDashboardAttributeFilter[]> =
    createSelector(selectFilterContextFilters, (filters): IDashboardAttributeFilter[] =>
        filters.filter(isDashboardAttributeFilter),
    );

/**
 * This selector returns dashboard's filter context date filter.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextDateFilter: DashboardSelector<IDashboardDateFilter | undefined> =
    createSelector(selectFilterContextFilters, (filters): IDashboardDateFilter | undefined =>
        filters.find(isDashboardCommonDateFilter),
    );

/**
 * This selector returns dashboard's filter context draggable filters.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextDraggableFilters: DashboardSelector<
    Array<IDashboardDateFilter | IDashboardAttributeFilter>
> = createSelector(
    selectFilterContextFilters,
    (filters): Array<IDashboardDateFilter | IDashboardAttributeFilter> =>
        filters.filter((f) => isDashboardDateFilterWithDimension(f) || isDashboardAttributeFilter(f)),
);

/**
 * This selector returns dashboard's filter context date filter with dimension specified.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextDateFiltersWithDimension: DashboardSelector<IDashboardDateFilter[]> =
    createSelector(selectFilterContextFilters, (filters): IDashboardDateFilter[] =>
        filters.filter(isDashboardDateFilterWithDimension),
    );

/**
 * Creates a selector for selecting date filter by its dataset {@link @gooddata/sdk-model#ObjRef}.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextDateFilterByDataSet: (
    dataSet: ObjRef,
) => (state: DashboardState) => IDashboardDateFilter | undefined = createMemoizedSelector((dataSet: ObjRef) =>
    createSelector(selectFilterContextDateFiltersWithDimension, (dateFilters) => {
        return dateFilters.find((filter) => areObjRefsEqual(filter.dateFilter.dataSet, dataSet));
    }),
);

/**
 * Creates a selector for selecting date filter's index by its dataset {@link @gooddata/sdk-model#ObjRef}.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextDateFilterIndexByDataSet: (
    dataSet: ObjRef,
) => (state: DashboardState) => number = createMemoizedSelector((dataSet: ObjRef) =>
    createSelector(selectFilterContextDateFiltersWithDimension, (dateFilters) => {
        return dateFilters.findIndex((filter) => areObjRefsEqual(filter.dateFilter.dataSet, dataSet));
    }),
);

/**
 * Creates a selector for selecting draggable filter's index by its ref:
 * dataSet ref for date filters {@link @gooddata/sdk-model#ObjRef}
 * localIdentifier for attribute filters
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextDraggableFilterIndexByRef: (
    ref: ObjRef | string,
) => (state: DashboardState) => number = createMemoizedSelector((ref: ObjRef | string) =>
    createSelector(selectFilterContextFilters, (filters) => {
        return filters.findIndex((filter) => {
            if (isDashboardDateFilterWithDimension(filter) && isObjRef(ref)) {
                return areObjRefsEqual(filter.dateFilter.dataSet, ref);
            }
            if (isDashboardAttributeFilter(filter) && typeof ref === "string") {
                return filter.attributeFilter.localIdentifier === ref;
            }
            return false;
        });
    }),
);

/**
 * Creates a selector for selecting attribute filter by its displayForm {@link @gooddata/sdk-model#ObjRef}.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextAttributeFilterByDisplayForm: (
    displayForm: ObjRef,
) => (state: DashboardState) => IDashboardAttributeFilter | undefined = createMemoizedSelector(
    (displayForm: ObjRef) =>
        createSelector(
            selectAttributeFilterDisplayFormsMap,
            selectFilterContextAttributeFilters,
            (attributeDisplayFormsMap, attributeFilters) => {
                const df = attributeDisplayFormsMap.get(displayForm);
                if (!df) {
                    return undefined;
                }
                // try matching both uri and id in case the type of ref is different from what is in the ref field
                return attributeFilters.find(
                    (filter) =>
                        areObjRefsEqual(filter.attributeFilter.displayForm, idRef(df.id, "displayForm")) ||
                        areObjRefsEqual(filter.attributeFilter.displayForm, uriRef(df.uri)),
                );
            },
        ),
);

/**
 * Creates a selector for selecting attribute filter by its localId.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextAttributeFilterByLocalId: (
    localId: string,
) => DashboardSelector<IDashboardAttributeFilter | undefined> = createMemoizedSelector((localId: string) =>
    createSelector(selectFilterContextAttributeFilters, (attributeFilters) =>
        attributeFilters.find((filter) => filter.attributeFilter.localIdentifier === localId),
    ),
);

/**
 * Creates a selector for selecting attribute filter index by its localId.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectFilterContextAttributeFilterIndexByLocalId: (
    localId: string,
) => DashboardSelector<number> = createMemoizedSelector((localId: string) =>
    createSelector(selectFilterContextAttributeFilters, (attributeFilters) =>
        attributeFilters.findIndex((filter) => filter.attributeFilter.localIdentifier === localId),
    ),
);

/**
 * Creates a selector for selecting all descendants of the attribute filter with given localId.
 *
 * @remarks
 * Invocations before initialization lead to invariant errors.
 * This selector should only be used when circular filter dependencies are not allowed.
 *
 * @public
 */
export const selectAttributeFilterDescendants: (localId: string) => DashboardSelector<string[]> =
    createMemoizedSelector((localId: string) =>
        createSelector(
            selectFilterContextAttributeFilters,
            selectSupportsCircularDependencyInFilters,
            (attributeFilters, supportsCircularDependencies) => {
                invariant(
                    !supportsCircularDependencies,
                    "Filter descendants cannot be computed as this backend supports circular dependencies in filters. Do not use this selector.",
                );

                const toCheck = compact([localId]);
                const result = new Set<string>();

                while (toCheck.length) {
                    const current = toCheck.pop();
                    const children = attributeFilters.filter((f) =>
                        f.attributeFilter.filterElementsBy?.some(
                            (parent) => parent.filterLocalIdentifier === current,
                        ),
                    );

                    children.forEach((child) => {
                        result.add(child.attributeFilter.localIdentifier!);
                        toCheck.push(child.attributeFilter.localIdentifier!);
                    });
                }

                return Array.from(result);
            },
        ),
    );

/**
 * Creates a selector for selecting all filters with different reference than the given one.
 *
 * @internal
 */
export const selectOtherContextAttributeFilters: (
    ref?: ObjRef,
) => DashboardSelector<IDashboardAttributeFilter[]> = createMemoizedSelector((ref?: ObjRef) =>
    createSelector(selectFilterContextAttributeFilters, (attributeFilters): IDashboardAttributeFilter[] => {
        return attributeFilters.filter(
            (attributeFilter) => !areObjRefsEqual(attributeFilter.attributeFilter.displayForm, ref),
        );
    }),
);

/**
 * Creates a selector to get a display form of the filter defined by its local identifier.
 *
 * @internal
 */
export const selectAttributeFilterDisplayFormByLocalId: (localId: string) => DashboardSelector<ObjRef> =
    createMemoizedSelector((localId: string) =>
        createSelector(selectFilterContextAttributeFilters, (filters) => {
            const filter = filters.find((filter) => filter.attributeFilter.localIdentifier === localId);

            invariant(filter, "Unable to find current filter to get its title.");

            return filter.attributeFilter.displayForm;
        }),
    );

/**
 * Creates a selector which checks for a circular dependency between filters.
 *
 * @remarks
 * This selector should only be used when circular filter dependencies are not allowed.
 *
 * @internal
 */
export const selectIsCircularDependency: (
    currentFilterLocalId: string,
    neighborFilterLocalId: string,
) => DashboardSelector<boolean> = createMemoizedSelector(
    (currentFilterLocalId: string, neighborFilterLocalId: string) =>
        createSelector(
            selectAttributeFilterDescendants(currentFilterLocalId),
            selectSupportsCircularDependencyInFilters,
            (descendants, supportsCircularDependencies) => {
                invariant(
                    !supportsCircularDependencies,
                    "No need to compute circular dependencies as this backend supports it in filters. Do not use this selector.",
                );

                return descendants.some((descendant) => descendant === neighborFilterLocalId);
            },
        ),
);

const MAX_DRAGGABLE_FILTERS_COUNT = 30;

/**
 * This selector returns whether any more attribute filters can be added.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 * @deprecated use selectCanAddMoreFilters instead
 */
export const selectCanAddMoreAttributeFilters: DashboardSelector<boolean> = createSelector(
    selectFilterContextAttributeFilters,
    selectFilterContextDateFiltersWithDimension,
    (attributeFilters, dateFiltersWithDimension) => {
        return attributeFilters.length + dateFiltersWithDimension.length < MAX_DRAGGABLE_FILTERS_COUNT;
    },
);

/**
 * This selector returns whether any more attribute filters can be added.
 *
 * @remarks
 * It is expected that the selector is called only after the filter context state is correctly initialized.
 * Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export const selectCanAddMoreFilters = selectCanAddMoreAttributeFilters;

/**
 * This selector returns information whether attribute filter from filter context has some dependencies.
 *
 * @internal
 */
export const selectIsAttributeFilterDependentByLocalIdentifier: (
    attributeFilterLocalIdentifier: string,
) => DashboardSelector<boolean> = createMemoizedSelector((attributeFilterLocalIdentifier: string) =>
    createSelector(
        selectFilterContextAttributeFilterByLocalId(attributeFilterLocalIdentifier),
        (filterContextAttributeFilter) => {
            return !isEmpty(filterContextAttributeFilter?.attributeFilter?.filterElementsBy);
        },
    ),
);
