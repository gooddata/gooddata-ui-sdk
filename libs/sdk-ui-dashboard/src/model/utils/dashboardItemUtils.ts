// (C) 2021-2022 GoodData Corporation
import {
    DashboardItemDefinition,
    ExtendedDashboardItem,
    InternalDashboardItemDefinition,
    isCustomWidgetDefinition,
} from "../types/layoutTypes.js";
import {
    idRef,
    ObjRef,
    IDashboardObjectIdentity,
    isKpiWidgetDefinition,
    isInsightWidgetDefinition,
    isInsightWidget,
    isDashboardLayoutItem,
} from "@gooddata/sdk-model";
import { v4 as uuidv4 } from "uuid";

export function extractInsightRefsFromItems(items: ReadonlyArray<ExtendedDashboardItem>): ObjRef[] {
    const result: ObjRef[] = [];

    items.forEach((item) => {
        if (isInsightWidget(item.widget)) {
            result.push(item.widget.insight);
        }
    });

    return result;
}

const TemporaryIdentityPrefix = "@@GDC.DASH.TEMP";
function generateTemporaryIdentityForWidget(): IDashboardObjectIdentity {
    const id = `${TemporaryIdentityPrefix}-${uuidv4()}`;

    return {
        ref: idRef(id),
        identifier: id,
        uri: id,
    };
}

/**
 * Tests whether the provided object is/has a temporary identity. A temporary identity is used for those
 * objects which are not yet persisted however need to be reference-able from the dashboard itself.
 *
 * @param obj - object to test
 * @internal
 */
export function isTemporaryIdentity(obj: IDashboardObjectIdentity): boolean {
    return obj.identifier.startsWith(TemporaryIdentityPrefix);
}

/**
 * Adds temporary identity to all insight and KPI widget definitions found within the provided dashboard item
 * definitions.
 *
 * Having an identity for each widget is important for the dashboard; many places both internal and through
 * public API for different use-cases rely on identification of widget by reference.
 *
 * This function will map the input items so that the result never contains bare insight widget definitions
 * or kpi widget definitions (which do not have the identity).
 *
 * The identity assigned to those widgets is temporary one. It is a specially formatted identifier combined
 * with UUID. You can use {@link isTemporaryIdentity} function to test whether object has temporary
 * identity. When the dashboard gets saved, this temporary identity will change to permanent one that
 * will be assigned by the backend.
 *
 * @param items - items to map, will not be modified
 * @returns new array with the necessary items mapped to widgets with temporary identity.
 * @internal
 */
export function addTemporaryIdentityToWidgets(
    items: ReadonlyArray<DashboardItemDefinition>,
): InternalDashboardItemDefinition[] {
    return items.map((item) => {
        if (!isDashboardLayoutItem(item)) {
            return item;
        }

        if (
            isInsightWidgetDefinition(item.widget) ||
            isKpiWidgetDefinition(item.widget) ||
            isCustomWidgetDefinition(item.widget)
        ) {
            const temporaryIdentity = generateTemporaryIdentityForWidget();

            return {
                ...item,
                widget: {
                    ...item.widget,
                    ...temporaryIdentity,
                },
            } as any;
        }

        return item;
    });
}
