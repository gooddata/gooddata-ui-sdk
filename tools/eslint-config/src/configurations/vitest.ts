// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
