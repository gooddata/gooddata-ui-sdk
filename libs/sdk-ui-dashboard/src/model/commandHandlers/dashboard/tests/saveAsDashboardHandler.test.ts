// (C) 2021-2024 GoodData Corporation
import { beforeEach, describe, it, expect } from "vitest";
import { DashboardTester, preloadedTesterFactory } from "../../../tests/DashboardTester.js";
import { addLayoutSection, saveDashboard, saveDashboardAs } from "../../../commands/index.js";
import { TestInsightItem } from "../../../tests/fixtures/Layout.fixtures.js";
import { TestCorrelation } from "../../../tests/fixtures/Dashboard.fixtures.js";
import { DashboardCopySaved, DashboardSaved } from "../../../events/index.js";
import { selectBasicLayout } from "../../../store/layout/layoutSelectors.js";
import { isTemporaryIdentity } from "../../../utils/dashboardItemUtils.js";
import { selectFilterContextIdentity } from "../../../store/filterContext/filterContextSelectors.js";
import { selectDashboardTitle, selectPersistedDashboard } from "../../../store/meta/metaSelectors.js";

import { SimpleDashboardIdentifier } from "../../../tests/fixtures/SimpleDashboard.fixtures.js";

describe("save as dashboard handler", () => {
    describe("for a new dashboard", () => {
        let Tester: DashboardTester;
        // each test starts with a new dashboard

        beforeEach(async () => {
            await preloadedTesterFactory(
                (tester) => {
                    Tester = tester;
                },
                undefined,
                {
                    backendConfig: {
                        useRefType: "id",
                    },
                },
            );
        });

        const TestDashboardTitle = "My Dashboard Copy";

        it("should save an existing dashboard as a copy and switch to copy", async () => {
            await Tester.dispatchAndWaitFor(
                addLayoutSection(0, {}, [TestInsightItem], false, TestCorrelation),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
            );

            const initialSaveEvent: DashboardSaved = await Tester.dispatchAndWaitFor(
                saveDashboard(),
                "GDC.DASH/EVT.SAVED",
            );

            expect(initialSaveEvent.payload.newDashboard).toEqual(true);
            const originalState = Tester.state();

            const event: DashboardCopySaved = await Tester.dispatchAndWaitFor(
                saveDashboardAs(TestDashboardTitle, true),
                "GDC.DASH/EVT.COPY_SAVED",
            );

            expect(event.payload.dashboard.ref).not.toEqual(initialSaveEvent.payload.dashboard.ref);
            expect(event.payload.dashboard.title).toEqual(TestDashboardTitle);
            expect(event.ctx.dashboardRef).toEqual(event.payload.dashboard.ref);
            const newState = Tester.state();

            const originalLayout = selectBasicLayout(originalState);
            const newLayout = selectBasicLayout(newState);

            expect(newLayout.sections[0].items[0].widget).not.toEqual(
                originalLayout.sections[0].items[0].widget,
            );
            expect(isTemporaryIdentity(newLayout.sections[0].items[0].widget!)).toEqual(false);

            expect(selectDashboardTitle(newState)).not.toEqual(selectDashboardTitle(originalState));
            expect(selectFilterContextIdentity(newState)).not.toEqual(
                selectFilterContextIdentity(originalState),
            );
            expect(selectPersistedDashboard(newState)?.ref).not.toEqual(
                selectPersistedDashboard(originalState)?.ref,
            );
        });

        it("should save an existing dashboard as a copy and leave current dashboard open", async () => {
            await Tester.dispatchAndWaitFor(
                addLayoutSection(0, {}, [TestInsightItem], false, TestCorrelation),
                "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
            );

            const initialSaveEvent: DashboardSaved = await Tester.dispatchAndWaitFor(
                saveDashboard(),
                "GDC.DASH/EVT.SAVED",
            );

            expect(initialSaveEvent.payload.newDashboard).toEqual(true);
            const originalState = Tester.state();

            const event: DashboardCopySaved = await Tester.dispatchAndWaitFor(
                saveDashboardAs(TestDashboardTitle, false),
                "GDC.DASH/EVT.COPY_SAVED",
            );

            expect(event.payload.dashboard.ref).not.toEqual(initialSaveEvent.payload.dashboard.ref);
            expect(event.payload.dashboard.title).toEqual(TestDashboardTitle);
            const newState = Tester.state();

            const originalLayout = selectBasicLayout(originalState);
            const newLayout = selectBasicLayout(newState);

            expect(newLayout.sections[0].items[0].widget).toEqual(originalLayout.sections[0].items[0].widget);
            expect(selectDashboardTitle(newState)).toEqual(selectDashboardTitle(originalState));
            expect(selectFilterContextIdentity(newState)).toEqual(selectFilterContextIdentity(originalState));
            expect(selectPersistedDashboard(newState)?.ref).toEqual(
                selectPersistedDashboard(originalState)?.ref,
            );
        });

        it("should emit correct events", async () => {
            await Tester.dispatchAndWaitFor(
                saveDashboardAs("My Dashboard Copy", false, true, TestCorrelation),
                "GDC.DASH/EVT.COPY_SAVED",
            );

            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });

    describe("for an existing dashboard", () => {
        let Tester: DashboardTester;

        beforeEach(async () => {
            await preloadedTesterFactory((tester) => {
                Tester = tester;
            }, SimpleDashboardIdentifier);
        });

        const TestDashboardTitle = "My Dashboard Copy";

        it("should save an existing emit correct events", async () => {
            await Tester.dispatchAndWaitFor(
                saveDashboardAs(TestDashboardTitle, true, undefined, TestCorrelation),
                "GDC.DASH/EVT.COPY_SAVED",
            );

            expect(Tester.emittedEventsDigest()).toMatchSnapshot();
        });
    });
});
