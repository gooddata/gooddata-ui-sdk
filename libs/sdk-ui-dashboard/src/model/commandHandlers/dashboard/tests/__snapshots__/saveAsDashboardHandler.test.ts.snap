// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`save as dashboard handler > for a new dashboard > should emit correct events 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.SAVEAS",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COPY_SAVED",
  },
]
`;

exports[`save as dashboard handler > for an existing dashboard > should save an existing emit correct events 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.SAVEAS",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COPY_SAVED",
  },
]
`;

exports[`save as dashboard handler > for an existing dashboard > should save an existing locked dashboard as a copy with isLocked flag cleaned 1`] = `
[
  {
    "commandType": "GDC.DASH/CMD.SAVEAS",
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COMMAND.STARTED",
  },
  {
    "correlationId": "testCorrelationId",
    "type": "GDC.DASH/EVT.COPY_SAVED",
  },
]
`;
